### DISCUSSION TITLE
How to check the file size before an upload in a `BEFORE TRIGGER FUNCTION`?

### DISCUSSION THREAD

**Discussion by bennik88** [supabase/supabase#14643] [Questions]
I want to protect uploads "serverside" in different ways and to do that I need the file size of the file the user wants to upload in the backend.
What is the cleanest way to check the file size? I thought of doing it in SQL with a trigger function as basically all of my protection and validation lies there.

But I was not able to retrieve the file size in any way before uploading. And there seems to also be a seconds problem to consider cause that it seems to upload before the file size is passed over as of the metadata is only written after the S3 upload. (https://github.com/orgs/supabase/discussions/6540)

Is there any parameter I'm missing to check the filesize of a file before uploading in my trigger function? Is there any `NEW` key that I'm missing I could access?

The trigger is straightforward:
```sql
-- Create trigger for function call on upload to user storage
CREATE TRIGGER is_upload_valid
BEFORE INSERT
ON storage.objects
FOR EACH ROW
EXECUTE PROCEDURE validate_upload();

CREATE OR REPLACE FUNCTION validate_upload()
RETURNS TRIGGER AS
$$
DECLARE
  file_size NUMERIC;
BEGIN
  -- How to access the file size here?
```

**Comment by GaryAustin1** [comment]
You are not missing anything.  There is no metadata in the insert to test RLS.  Only after successful s3 store.
I've not looked at Supabase's v3 storage (by request currently) source in detail but in a quick glance I don't think the process is changing from what was at that link.

Not sure you can do this with a trigger.

**Reply by bennik88** [reply]
@GaryAustin1 Thank you for engaging here. How would you then approach adding a security layer to check the file size yourself before uploading?
In the meantime, I was trying to use the new Supabase AI helper, which gave me some promising information. But I think the provided information is not correct or existing. I tried some of them. See attachment.
![Bildschirmfoto 2023-05-28 um 19 21 49](https://github.com/supabase/supabase/assets/62398136/2a7148f6-ee60-45d9-a1c1-7a87152fcda9)

It also gave me other solutions but I think most of them are not working.

**Reply by GaryAustin1** [reply]
I would not rely on anything about storage with the various AI's.   If you delete a row in the storage.objects table for instance, the s3 data is orphaned forever.  This has been suggested by at least two AI's (probably both chatGPT).  There is no way to directly do storage from SQL.  You have to use an http or pg-net extension to send out http requests to the REST API.

You might look at using an edge function to do your file uploads...
https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/file-upload-storage/index.ts
